### PR TITLE
:triangular_flag_on_post: Add vanilla LO-RANSAC

### PR DESCRIPTION
**Pull request by ducha-aiki**
#### Changes
Added (parallel) RANSAC implementation, which does sampling in batches.
ToDo: documentation and tests.

After we add this, to have state-of-the-art RANSAC, we will need to implement the following:

- [7-pt solver](https://github.com/opencv/opencv/blob/master/modules/calib3d/src/fundam.cpp#L498) for fundamental matrix and [5-pt solver](https://github.com/opencv/opencv/blob/master/modules/calib3d/src/five-point.cpp#L42) for essential matrix. That is top priority
- sota polisher, like [MAGSAC++](https://openaccess.thecvf.com/content_CVPR_2020/html/Barath_MAGSAC_a_Fast_Reliable_and_Accurate_Robust_Estimator_CVPR_2020_paper.html) Second priotity
- [GC-RANSAC](https://cmp.felk.cvut.cz/~matas/papers/barath-2018-gc_ransac-cvpr.pdf) as inlier/outlier classifier
- various degeneracy checks

To discuss:  

1. Do we want to have batched input? Given that we will use batching internally, I am not sure about simplicity of its support
2. Do we want dict as an input, or just tensors? I more tend to tensors

### CODE REVIEW

File: kornia/geometry/ransac.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    find_homography_dlt_iterated,
+    symmetrical_epipolar_distance,
+)
+from kornia.geometry.homography import symmetric_transfer_error
+
+__all__ = ["RANSAC"]
+
+
+class RANSAC(nn.Module):
+    """Module for robust geometry estimation with RANSAC. https://en.wikipedia.org/wiki/Random_sample_consensus.
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:21]:
```suggestion
    """Module for robust geometry estimation with RANSAC.
    
    https://en.wikipedia.org/wiki/Random_sample_consensus
```

File: kornia/geometry/ransac.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+    Args:
+        model_type (str): type of model to estimate, e.g. "homography".
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:24]:
```suggestion
        model_type: type of model to estimate, e.g. "homography" or "fundamental".
```

File: kornia/geometry/ransac.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        batch_size (int): number of generated samples at once .
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:25]:
```suggestion
        batch_size: number of generated samples at once.
```

File: kornia/geometry/ransac.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        max_iter (int): maximum batches to generate. Actual number of models to try is batch_size * max_iter.
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:26]:
```suggestion
        max_iterations: maximum batches to generate. Actual number of models to try is ``batch_size * max_iterations``.
```

File: kornia/geometry/ransac.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        inl_th (float): threshold for the correspondence to be an inlier
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:27]:
```suggestion
        inliers_threshold: threshold for the correspondence to be an inlier.
```

File: kornia/geometry/ransac.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        confidence (float): desired confidence of the result, used for the early stopping.
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:28]:
```suggestion
        confidence: desired confidence of the result, used for the early stopping.
```

File: kornia/geometry/ransac.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+        max_lo_iters (int): number of local optimization (polishing) iterations
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:29]:
```suggestion
        max_local_iterations: number of local optimization (polishing) iterations.
```

File: kornia/geometry/ransac.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+    def sample(self, sample_size: int,
+               pop_size: int,
+               batch_size: int, device=torch.device('cpu')) -> torch.Tensor:
+        """Minimal sampler, but unlike traditional RANSAC we sample in batches to get benefit of the parallel
+        processing, esp.
+
+        on GPU
+        """
+        out: torch.Tensor = torch.empty(batch_size, sample_size)
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:75]:
```suggestion
        out: torch.Tensor = torch.empty(batch_size, sample_size, device=device, dtype=torch.int32)
```

File: kornia/geometry/ransac.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+        # for loop, until https://github.com/pytorch/pytorch/issues/42502 accepted
+        for i in range(batch_size):
+            out[i] = torch.randperm(pop_size, dtype=torch.int32, device=device)[:sample_size]
+        return out
+
+    @staticmethod
+    def max_samples_by_conf(n_inl: int, num_tc: int, sample_size: int, conf: float):
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:82]:
```suggestion
    def max_samples_by_conf(n_inl: int, num_tc: int, sample_size: int, conf: float) -> float:
```

File: kornia/geometry/ransac.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+        if len(kp1.shape) == 2:
+            kp1 = kp1[None]
+        if len(kp2.shape) == 2:
+            kp2 = kp2[None]
+        batch_size = models.shape[0]
+        errors = self.error_fn(kp1.expand(batch_size, -1, 2),
+                               kp2.expand(batch_size, -1, 2),
+                               models)
+        inl = (errors <= inl_th)
+        models_score = inl.float().sum(dim=1)
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:114]:
```suggestion
        models_score = inl.to(kp1).sum(dim=1)
```

File: kornia/geometry/ransac.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        # ToDo: add (model-specific) verification of the samples,
+        # E.g. contraints on not to be a degenerate sample
+        return kp1, kp2
+
+    def remove_bad_models(self, models: torch.Tensor) -> torch.Tensor:
+        # ToDo: add more and better degenerate model rejection
+        # For now it is simple and hardcoded
+        main_diagonal = torch.diagonal(models,
+                                       dim1=1,
+                                       dim2=2)
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:132]:
```suggestion
                                       dim2=2, device=self.device)
```
maybe we should store device

File: kornia/geometry/ransac.py [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+                                                num_inl,
+                                                dtype=kp1_inl.dtype,
+                                                device=kp1_inl.device))
+        return model
+
+    def forward(self,
+                kp1: torch.Tensor,
+                kp2: torch.Tensor,
+                weights: Optional[torch.Tensor] = None) -> Tuple[torch.Tensor, torch.Tensor]:
+        '''
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:156]:
```suggestion
        """Main forward method to execute the RANSAC algorithm.
```

File: kornia/geometry/ransac.py [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+            Args:
+                kp1 (torch.Tensor): source image keypoints :math:`(N, 2)`.
+                kp2 (torch.Tensor): distance image keypoints :math:`(N, 2)`.
+                weights (torch.Tensor): optional correspondences weights. Not used now
+
+            Returns:
+                - estimated model, shape of :math:`(1, 3, 3)`.
+                - the inlier/outlier mask, shape of :math:`(1, N), where N is number of input correspondences`.'''
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:164]:
```suggestion
                - the inlier/outlier mask, shape of :math:`(1, N), where N is number of input correspondences`.
            """
```

File: kornia/geometry/ransac.py [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+            raise ValueError(f"Invalid kp1 shape, we expect Nx2 Got: {kp1.shape}")
+        if not len(kp2.shape) == 2:
+            raise ValueError(f"Invalid kp2 shape, we expect Nx2 Got: {kp2.shape}")
+        if not (kp1.shape[0] == kp2.shape[0]) or (kp1.shape[0] < self.minimal_sample_size):
+            raise ValueError(f"kp1 and kp2 should be \
+                             equal shape at at least [{self.minimal_sample_size}, 2], \
+                             got {kp1.shape}, {kp2.shape}")
+
+        best_score_total: float = 1.0
+        num_tc: int = len(kp1)
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:179]: what `tc` stands for ? find a more explicit name for that variable

File: kornia/geometry/ransac.py [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+        best_model_total = torch.zeros(3, 3, dtype=kp1.dtype, device=kp1.device)
+        inliers_best_total: torch.Tensor = torch.zeros(num_tc, 1, device=kp1.device).bool()
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:181]:
```suggestion
        inliers_best_total: torch.Tensor = torch.zeros(num_tc, 1, device=kp1.device, dtype=torch.bool)
```

File: kornia/geometry/ransac.py [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+        for i in range(self.max_iter):
+            # Sample minimal samples in batch to estimate models
+            idxs = self.sample(self.minimal_sample_size, num_tc, self.batch_size).long()
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:184]:
```suggestion
            idxs = self.sample(self.minimal_sample_size, num_tc, self.batch_size)
```
`sample` returns already ints

File: kornia/geometry/ransac.py [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+    Args:
+        model_type (str): type of model to estimate, e.g. "homography".
+        batch_size (int): number of generated samples at once .
+        max_iter (int): maximum batches to generate. Actual number of models to try is batch_size * max_iter.
+        inl_th (float): threshold for the correspondence to be an inlier
+        confidence (float): desired confidence of the result, used for the early stopping.
+        max_lo_iters (int): number of local optimization (polishing) iterations
+
+    Returns:
+        - estimated model, shape of :math:`(1, 3, 3)`.
```
Comment (edgarriba) [inline kornia/geometry/ransac.py:32]:
```suggestion
        - estimated model, shape of :math:`(3, 3)`.
```
in the tests you don't return batch. I guess doesn't make sense here batch

File: test/geometry/test_ransac.py [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+from kornia.testing import assert_close
+
+
+class TestRANSACHomography:
+    def test_smoke(self, device, dtype):
+        points1 = torch.rand(4, 2, device=device, dtype=dtype)
+        points2 = torch.rand(4, 2, device=device, dtype=dtype)
+        ransac = RANSAC('homography').to(device=device, dtype=dtype)
+        H, inliers = ransac(points1, points2)
+        assert H.shape == (3, 3)
```
Comment (edgarriba) [inline test/geometry/test_ransac.py:19]: lovely example/snippet :)

File: test/geometry/test_ransac.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+    @pytest.mark.xfail(reason="might slightly and randomly imprecise due to RANSAC randomness")
+    def test_dirty_points(self, device, dtype):
+        # generate input data
+
+        H = torch.eye(3, dtype=dtype, device=device)
+        H[:2] = H[:2] + 0.1 * torch.rand_like(H[:2])
+        H[2:, :2] = H[2:, :2] + 0.001 * torch.rand_like(H[2:, :2])
+
+        points_src = 100.0 * torch.rand(1, 20, 2, device=device, dtype=dtype)
+        points_dst = kornia.transform_points(H[None], points_src)
```
Comment (edgarriba) [inline test/geometry/test_ransac.py:30]:
```suggestion
        points_dst = kornia.geometry.transform_points(H[None], points_src)
```

File: test/geometry/test_ransac.py [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+
+        # making last point an outlier
+        points_dst[:, -1, :] += 800
+        ransac = RANSAC('homography', inl_th=0.5, max_iter=20).to(device=device, dtype=dtype)
+        # compute transform from source to target
+        dst_homo_src, inliers = ransac(points_src[0], points_dst[0])
+
+        assert_close(
+            kornia.transform_points(dst_homo_src[None], points_src[:, :-1]),
```
Comment (edgarriba) [inline test/geometry/test_ransac.py:39]:
```suggestion
            kornia.geometry.transform_points(dst_homo_src[None], points_src[:, :-1]),
```

File: test/geometry/test_ransac.py [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+import torch
+from torch.autograd import gradcheck
+
+import kornia
+import kornia.testing as utils
+from kornia.geometry import RANSAC
+from kornia.testing import assert_close
+
+
+class TestRANSACHomography:
```
Comment (edgarriba) [inline test/geometry/test_ransac.py:13]: add a value tests as in the fundamental. Check if jit, onnx, (gradcheck?) work here

**Review by edgarriba — CHANGES_REQUESTED** [review]

File: kornia/geometry/ransac.py (hunk 14 above)
Comment (ducha-aiki) [reply kornia/geometry/ransac.py:179]: tentative correspondences

File: kornia/geometry/ransac.py (hunk 16 above)
Comment (ducha-aiki) [reply kornia/geometry/ransac.py:184]: No, it doesn't. dtype=torch.int32 is not torch.long and cannot be used for indexing

File: test/geometry/test_ransac.py (hunk 21 above)
Comment (ducha-aiki) [reply test/geometry/test_ransac.py:13]: Why? We already have numeric test in dirty_points

Comment (ducha-aiki) [reply test/geometry/test_ransac.py:13]: Will add jit thing

**Comment by ducha-aiki** [comment]
Future work:

profile and speed-up model estimation (find_homography_flt) and find_fundamental functions, as they are now dominating the RANSAC time. Then comes (5x faster that model estimation) model verification.

**Pull request merged**